In [1]:
import pandas as pd
from bs4 import BeautifulSoup
import requests

<h1>Составление талбицы с CPI</h1>

In [2]:
url = "https://www.usinflationcalculator.com/inflation/consumer-price-index-and-annual-percent-changes-from-1913-to-2008/"

In [3]:
r = requests.get(url)
data = r.text
soup = BeautifulSoup(data, "html.parser")

In [4]:
table = soup.find('table')
rows = table.tbody.findAll('tr')

In [5]:
years = list()
cpis = list()

In [6]:
for row in rows:
    year = row.findAll('td')[0].get_text()
    if year.isdigit() and int(year) <= 2017:
        years.append(int(year))
        cpis.append(float(row.findAll('td')[13].get_text()))

In [7]:
cpi_table = pd.DataFrame({
    "year": years,
    "avg_annual_cpi": cpis
})

<h1>Запись таблицы c CPI в файл</h1>

In [8]:
cpi_table.to_csv("../data/Lab1_Koren_CPI.csv", sep=',')

In [9]:
def get_real_value(nominal_amt, old_cpi, new_cpi):
    real_value = (nominal_amt * new_cpi) / old_cpi
    return real_value

In [10]:
CPI_2017 = float(cpi_table[cpi_table['year'] == 2017]['avg_annual_cpi'])

C:\Users\qwerty\AppData\Local\Temp\ipykernel_10676\793858478.py:1: FutureWarning: Calling float on a single element Series is deprecated and will raise a TypeError in the future. Use float(ser.iloc[0]) instead
  CPI_2017 = float(cpi_table[cpi_table['year'] == 2017]['avg_annual_cpi'])


<h1></h1>

<h1>Считываем данных из csv файла, где хранится информация о фильмах</h1>

In [11]:
movies_table = pd.read_csv('../data/Dataset1.csv')

In [12]:
movies_table.head(10)

color      Director_Name  num_Critic_for_reviews  duration  \
0   Color      James Cameron                   723.0     178.0   
1  Colour     Gore Verbinski                   302.0     169.0   
2  Colour         Sam Mendes                   602.0     148.0   
3   Color  Christopher Nolan                   813.0     164.0   
4     NaN        Doug Walker                     NaN       NaN   
5  Colour     Andrew Stanton                   462.0     132.0   
6   Color          Sam Raimi                   392.0     156.0   
7   Color       Nathan Greno                   324.0     100.0   
8  Colour        Joss Whedon                   635.0     141.0   
9  Colour        David Yates                   375.0     153.0   

   director_Facebook_likes  actor_3_Facebook_likes       actor_2_name  \
0                      0.0                   855.0   Joel David Moore   
1                    563.0                  1000.0      Orlando Bloom   
2                      0.0                   161.0       Rory Kinnear   
3                  22000.0                 23000.0     Christian Bale   
4                    131.0                     NaN         Rob Walker   
5                    475.0                   530.0    Samantha Morton   
6                      0.0                  4000.0       James Franco   
7                     15.0                   284.0       Donna Murphy   
8                      0.0                 19000.0  Robert Downey Jr.   
9                    282.0                 10000.0   Daniel Radcliffe   

   Actor_1_Facebook_likes        gross  \
0                  1000.0  760505847.0   
1                 40000.0  309404152.0   
2                 11000.0  200074175.0   
3                 27000.0  448130642.0   
4                   131.0          NaN   
5                   640.0   73058679.0   
6                 24000.0  336530303.0   
7                   799.0  200807262.0   
8                 26000.0  458991599.0   
9                 25000.0  301956980.0   

                                              genres  ...  \
0                    Action|Adventure|Fantasy|Sci-Fi  ...   
1                           Action|Adventure|Fantasy  ...   
2                          Action|Adventure|Thriller  ...   
3                                    Action|Thriller  ...   
4                                        Documentary  ...   
5                            Action|Adventure|Sci-Fi  ...   
6                           Action|Adventure|Romance  ...   
7  Adventure|Animation|Comedy|Family|Fantasy|Musi...  ...   
8                            Action|Adventure|Sci-Fi  ...   
9                   Adventure|Family|Fantasy|Mystery  ...   

  num_user_for_reviews language  country  content_rating       budget  \
0               3054.0  English      USA           PG-13  237000000.0   
1               1238.0  English      USA           PG-13  300000000.0   
2                994.0  English       UK           PG-13  245000000.0   
3               2701.0  English      USA           PG-13  250000000.0   
4                  NaN      NaN      NaN             NaN          NaN   
5                738.0  English      USA           PG-13  263700000.0   
6               1902.0  English      USA           PG-13  258000000.0   
7                387.0  English      USA              PG  260000000.0   
8               1117.0  English      USA           PG-13  250000000.0   
9                973.0  English       UK              PG  250000000.0   

   title_year actor_2_facebook_likes imdb_score  aspect_ratio  \
0      2009.0                  936.0        7.9          1.78   
1      2007.0                 5000.0        7.1          2.35   
2      2015.0                  393.0        6.8          2.35   
3      2012.0                23000.0        8.5          2.35   
4         NaN                   12.0        7.1           NaN   
5      2012.0                  632.0        6.6          2.35   
6      2007.0                11000.0        6.2          2.35   
7      2010.0 

<h1></h1>

<h1>Прописываем в пустые ячейки "unknown"</h1>

In [13]:
movies_table.isna().sum()

color                          18
Director_Name                 171
num_Critic_for_reviews        116
duration                       84
director_Facebook_likes       171
actor_3_Facebook_likes         90
actor_2_name                   80
Actor_1_Facebook_likes         75
gross                         939
genres                         69
actor_1_name                   75
movie_Title                    69
num_voted_users                69
 cast_total_facebook_likes     69
actor_3_name                   90
facenumber_in_poster           82
plot_keywords                 216
movie_imdb_link                69
num_user_for_reviews           87
language                       82
country                        72
content_rating                366
budget                        549
title_year                    174
actor_2_facebook_likes         80
imdb_score                     69
aspect_ratio                  389
movie_facebook_likes;          69
dtype: int64

In [14]:
columns_to_replace =['title_year', 'gross', 'budget']

for row in columns_to_replace:
    movies_table[row] = movies_table[row].fillna('unknown')

In [15]:
movies_table.isna().sum()

color                          18
Director_Name                 171
num_Critic_for_reviews        116
duration                       84
director_Facebook_likes       171
actor_3_Facebook_likes         90
actor_2_name                   80
Actor_1_Facebook_likes         75
gross                           0
genres                         69
actor_1_name                   75
movie_Title                    69
num_voted_users                69
 cast_total_facebook_likes     69
actor_3_name                   90
facenumber_in_poster           82
plot_keywords                 216
movie_imdb_link                69
num_user_for_reviews           87
language                       82
country                        72
content_rating                366
budget                          0
title_year                      0
actor_2_facebook_likes         80
imdb_score                     69
aspect_ratio                  389
movie_facebook_likes;          69
dtype: int64

<h1></h1>

<h1>Вычисление валовой и бюджетной стоимости в реальной сумме</h1>

In [16]:
real_domestic_gross = []
real_budget_values = []

In [17]:
for index, row in movies_table.iterrows():
    gross = row['gross']
    budget = row['budget']
    year = row['title_year']
    if year == "unknown"  or gross == "unknown" or budget == "unknown":
        real_domestic_gross.append(gross)
        real_budget_values.append(budget)
        continue
    cpi = float(cpi_table[cpi_table['year'] == int(year)]['avg_annual_cpi'])
    
    real_gross = get_real_value(gross, cpi, CPI_2017)
    real_budget = get_real_value(budget, cpi, CPI_2017)
    real_domestic_gross.append(real_gross)
    real_budget_values.append(real_budget)
movies_table["real_domestic_gross"] = real_domestic_gross
movies_table["real_budget"] = real_budget_values

C:\Users\qwerty\AppData\Local\Temp\ipykernel_10676\1815066857.py:9: FutureWarning: Calling float on a single element Series is deprecated and will raise a TypeError in the future. Use float(ser.iloc[0]) instead
  cpi = float(cpi_table[cpi_table['year'] == int(year)]['avg_annual_cpi'])


<h1></h1>

<h1>Вычисление ROI и абсолютной прибыли</h1>

In [18]:
profits = []
roi_vals = []

In [19]:
for index, row in movies_table.iterrows():
    if row["real_domestic_gross"] == "unknown" or row['real_budget'] == "unknown":
        profits.append("unknown")
        roi_vals.append("unknown")
        continue
    profit = row["real_domestic_gross"] - row['real_budget']
    budget = row['real_budget']
    num = profit - budget
    den = budget
    roi = (num / den) * 100
    
    profits.append(profit)
    roi_vals.append(roi)

movies_table['profit'] = profits
movies_table['roi'] = roi_vals

<h1></h1>

<h1>Удаление дубликатов</h1>

In [20]:
movies_table.duplicated().sum()

45

In [21]:
movies_table = movies_table.drop_duplicates().reset_index(drop=True)

In [22]:
movies_table.duplicated().sum()

0

<h1></h1>

<h1>Гипотезы</h1>

1. Большая часть фильмов окупилась 
2. Action - самый чаcтый жанр
3. Средняя оценка фильмов 6.5 +- 0.2


<h2>Проверка первой гипотезы</h2>

In [23]:
count_profit_films = 0
count_films = 0
for index, row in movies_table.iterrows():
    if row['profit'] == "unknown":
        continue
    if row['profit'] > 0:
        count_profit_films += 1
    count_films += 1

print((count_profit_films * 100) / count_films)

52.584623458409865


Из волученного результата видно, что 52.58% фильмов окупились.

<h2>Проверка второй гипотезы</h2>

In [24]:
popularity_of_genres = dict()
for index, row in movies_table.iterrows():
    if not isinstance(row['genres'], str):
        continue
    for genre in row['genres'].split('|'):
        if genre not in popularity_of_genres:
            popularity_of_genres[genre] = 1
        else:
            popularity_of_genres[genre] += 1

popularity_of_genres = pd.DataFrame.from_dict(popularity_of_genres, orient='index', columns=['popularity'])
        

In [25]:
popularity_of_genres = popularity_of_genres.sort_values(by=['popularity'], ascending=False)
popularity_of_genres.head()

popularity
Drama           2543
Comedy          1826
Thriller        1389
Action          1134
Romance         1081

По таблице видно, что гипотеза оказалась ошибочной, и самый частый жанр - Drama

<h2>Проверка третий гипотезы</h2>

In [26]:
print(movies_table['imdb_score'].mean(numeric_only=True))

6.438498985801217


Средняя оценка за фильм 6.4. Гипотеза подтверждена.

<h1>Доп. Задание</h1>

In [ ]:
pivot_table = df.pivot_table(index=['city', 'genre'], aggfunc='count', values='track')